tests/tools/app_store_connect/test_publish_action.py

```diff
@@ -65,6 +65,107 @@ def test_publish_action_with_username_and_password(_mock_altool, namespace_kwarg
+            **localization_arguments,
+        )
+
+    assert str(error_info.value) \
+        == 'Both --locale and --whats-new need to be defined for providing beta build localization'
+
+
+def test_publish_action_with_localization_and_no_testflight_submission(namespace_kwargs):
+    namespace_kwargs.update({'action': 'publish'})
+    cli_args = argparse.Namespace(**namespace_kwargs)
+    patterns = [pathlib.Path('path.pattern')]
+
+    with pytest.raises(IOError) as error_info:
+        AppStoreConnect.from_cli_args(cli_args).publish(
+            application_package_path_patterns=patterns,
+            apple_id='name@example.com',
+            app_specific_password=Types.AppSpecificPassword('xxxx-yyyy-zzzz-wwww'),
+            locale='en-GB',
+            whats_new="What's new",
+        )
+
+    assert str(error_info.value) \
+        == '--testflight flag is required for providing beta build localization'
+
+
+def test_publish_action_testflight_with_localization(namespace_kwargs):
+    class MockBuild:
+        def __init__(self, build_id):
+            self.id = build_id
+
+    class MockIpa:
+        def __init__(self, path: pathlib.Path):
+            self.path = path
+
+        @classmethod
+        def get_text_summary(cls):
+            return ''
+
+    namespace_kwargs.update({'action': 'publish'})
+
+    cli_args = argparse.Namespace(**namespace_kwargs)
+    with mock.patch.object(AppStoreConnect, 'find_paths') as mock_find_paths, \
+            mock.patch.object(AppStoreConnect, '_get_publishing_application_packages') as mock_get_packages, \
+            mock.patch.object(AppStoreConnect, '_upload_artifact_with_altool') as mock_upload, \
+            mock.patch.object(AppStoreConnect, '_validate_artifact_with_altool') as mock_validate, \
+            mock.patch.object(AppStoreConnect, '_get_uploaded_build') as mock_get_build, \
+            mock.patch.object(AppStoreConnect, 'create_beta_app_review_submission') as mock_create_review, \
+            mock.patch.object(AppStoreConnect, 'create_beta_build_localization') as mock_create_localization:
+        ipa_path = pathlib.Path('app.ipa')
+        mock_find_paths.return_value = [ipa_path]
+        mock_get_packages.return_value = [MockIpa(ipa_path)]
+        build = MockBuild('1525e3c9-3015-407a-9ba5-9addd2558224')
+        mock_get_build.return_value = [build, '1.0.0']
+
+        patterns = [pathlib.Path('path.pattern')]
+        AppStoreConnect.from_cli_args(cli_args).publish(
+            application_package_path_patterns=patterns,
+            apple_id='name@example.com',
+            app_specific_password=Types.AppSpecificPassword('xxxx-yyyy-zzzz-wwww'),
+            submit_to_testflight=True,
+            locale='en-GB',
+            whats_new="What's new",
+        )
+        mock_validate.assert_called()
+        mock_upload.assert_called()
+        mock_create_review.assert_called_with(build.id)
+        mock_get_packages.assert_called_with(patterns)
```

Comment:
I feel like this validation should be the first. It is easier to follow things in their natural occuring order.